Add raw int measurement value type (#354)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

Integrity Registers can be used to model the PCRs in a TPM or vTPM, in which case the identifier is the register index, or other kinds of vendor-specific measured objects.

> 

Author: @andrew-draper requests this PR to contain a matching algorithm.

Author: Also example diags.

##### Linear Privilege Level {#sec-comid-linear-privlevel}

A Linear Privilege Level describes an operating privilege for the target environment.
The convention is that 0 is the highest privilege, and higher numbers correspond to fewer privileges.
The semantics of the ordering indicate a total inclusion of all higher privilege levels.
This means that there is no privilege strictly afforded to, e.g., privilege level 3 that isn't also available to privilege level 1.

~~~ cddl
{::include cddl/linear-privilege-level.cddl}
~~~

> For environments that count privilege in the opposite order, it is recommended to represent the privilege levels with non-positive numbers, where 0 is the lowest privilege, and some negative integer is the highest.

Author: ```suggestion
For environments that count privilege in the opposite order, it is recommended to represent the privilege levels with non-positive numbers, where 0 is the lowest privilege, and some negative integer is the highest.
The full range if signed integers may be used.
```

Integrity Registers can be used to model the PCRs in a TPM or vTPM, in which case the identifier is the register index, or other kinds of vendor-specific measured objects.

##### Linear Privilege Level {#sec-comid-linear-privlevel}

> A Linear Privilege Level describes an operating privilege for the target environment.

Reviewer: I would swap the bottom sentence to state that:

Author: Please clarify.

> A Linear Privilege Level describes an operating privilege for the target environment.

Reviewer: ```suggestion
A Linear Privilege Level describes an operating privilege for the target environment. The privilege can be expressed in either positive integer ranges or in negative integer ranges.
```

> The convention is that 0 is the highest privilege, and higher numbers correspond to fewer privileges.

Reviewer: ```suggestion
When expressed in positive integer ranges the convention is that 0 is the highest privilege, and higher numbers correspond to fewer privileges. No negative number representation is allowed, in this setting.
```

Author: What setting? This is convention.

Author: I don't understand this comment. Replace what with convention? I'm trying to encode a range comparison that requires an internalized notion of order.

Integrity Registers can be used to model the PCRs in a TPM or vTPM, in which case the identifier is the register index, or other kinds of vendor-specific measured objects.

##### Linear Privilege Level {#sec-comid-linear-privlevel}

A Linear Privilege Level describes an operating privilege for the target environment.
A privilege level is represented by a signed integer.
The semantics of the ordering indicate a total inclusion of all higher privilege levels.
This means that there is no privilege strictly afforded to, e.g., privilege level 3 that isn't also available to privilege level 1.

~~~ cddl
{::include cddl/linear-privilege-level.cddl}
~~~

The convention is that 0 is the highest privilege, and higher numbers correspond to fewer privileges.
For environments that count privilege in the opposite order, it is recommended to represent the privilege levels with non-positive numbers, where 0 is the lowest privilege, and some negative integer is the highest.

> The full range if signed integers may be used.

Reviewer: ```suggestion
The full range of signed integers may be used.
```

Integrity Registers can be used to model the PCRs in a TPM or vTPM, in which case the identifier is the register index, or other kinds of vendor-specific measured objects.

##### Raw Integral {#sec-comid-raw-integral}

A raw integral describes an integral value that can be compared with linear order in the target environment.

> A raw integral is represented with either major type 0 or major type 1 ints, or even tagged bignums.

Reviewer 2: What is the use case for bignums?

Reviewer 3: I have this question also. For example, when would someone use bigint vs. raw-value (which is bytes also) assuming a future version of raw-value could introduce a range of raw-value?

Instead, if an entry is found, the digest comparison proceeds as defined in {{sec-cmp-digests}} after equivalence has been found according to {{sec-comid-integrity-registers}}.
Note that it is not required for all the entries in the candidate entry to be used during matching: the condition ECT could consist of a subset of the device's register space. In TPM parlance, a TPM "quote" may report all PCRs in Evidence, while a condition ECT could describe a subset of PCRs.

##### Comparison for linear-privlevel entries

The ACS entry value stored under `measurement-values-map` codepoint 15 is a linear privilege level, which must have type `linear-privlevel-type-choice`.

> If the entry `linear-privlevel-type-choice` is an `int` or an `int` tagged with #6.564, then comparison with the `int` named as PRIV is as follows.

Reviewer: I am not clear when was PRIV introduced and what its context is? Can you please clarify..

Author: PRIV is referenced below when describing the comparison. PRIV here represents an exact privilege level in the ACS that will be compared against a CoRIM. I'm following the same wording form as for SVN.

> *  If the condition ECT value for `measurement-values-map` codepoint 15 is an untagged `int` or an `int` tagged with #6.564 then an equality comparison is performed on the `int` components.

Reviewer: When would one use `int` and whan would one use `tagged-int`. can you please explain as this is the base spec!

Author: Fixed.

The comparison MUST return true if the value of PRIV is equal to the `int` value in the condition ECT.

*  If the condition ECT value for `measurement-values-map` codepoint 15 is an `int-range` or an `int-range` tagged with #6.565 then a range inclusion comparison is performed.

> The comparison MUST return true if the value of PRIV is greater than or equal to the `min` value in the condition ECT AND the value of PRIV is less or equal to than the `max` value in the condition ECT

Reviewer 3: ```suggestion
The comparison MUST return true if the value of PRIV is greater than or equal to the `min` value in the condition ECT AND the value of PRIV is less than or equal to than the `max` value in the condition ECT
```

> If the entry `linear-privilege-type-choice` is an `int-range` or `int-range` tagged with #6.564, then comparison with the pair of `inf-int` values MINPRIV and MAXPRIV is as follows.

Reviewer 3: `inf-int` doesn't appear to be defined anywhere.

Author: Fixed.

##### Comparison for linear-privlevel entries

The ACS entry value stored under `measurement-values-map` codepoint 15 is a linear privilege level, which must have type `linear-privlevel-type-choice`.

If the entry `linear-privlevel-type-choice` is an `int`, name it PRIV and compare as follows.

*  If the condition ECT value for `measurement-values-map` codepoint 15 is an `int` then an equality comparison is performed on the `int` components.
The comparison MUST return true if the value of PRIV is equal to the `int` value in the condition ECT.

*  If the condition ECT value for `measurement-values-map` codepoint 15 is an `int-range` tagged with #6.564 then a range inclusion comparison is performed.
The comparison MUST return true if the value of PRIV is greater than or equal to the `min` value in the condition ECT AND the value of PRIV is less than or equal to than the `max` value in the condition ECT

> If the entry `linear-privilege-type-choice` is an `int-range` or `int-range` tagged with #6.564, then comparison with the pair of `opt-int` values MINPRIV and MAXPRIV is as follows.

Reviewer: I find it hard to understand the logic for 2539 beneath.. Is there a way to simplify/clarify it better??

> *  If the condition ECT value for `measurement-values-map` codepoint 15 is an `int` then the comparison MUST return true if and only if MINPRIV and MAXPRIV are equal and finite.

Reviewer: What is MINPRIV and MAXPRIV here, please set the context, like you did for PRIV Above.

Just say they are constant with Value X or Value Y.

Reviewer 3: If I reread the text above using MAXINT/MININT the wording seems confusing in that max infinity can't be equal to min infinity and infinity can be equal to a finite number at the same time. Or at least I have to be a theoretical mathematition to make it make sense. 

I think the text is trying to say that if the input value is type 'int' then the comparison must exactly match a value that is also of type 'int'. It is noted that MAXINT and MININT are special case values of type 'int' because we don't want to encode large numbers.

Author: I've changed the wording. Please take a look.

Instead, if an entry is found, the digest comparison proceeds as defined in {{sec-cmp-digests}} after equivalence has been found according to {{sec-comid-integrity-registers}}.
Note that it is not required for all the entries in the candidate entry to be used during matching: the condition ECT could consist of a subset of the device's register space. In TPM parlance, a TPM "quote" may report all PCRs in Evidence, while a condition ECT could describe a subset of PCRs.

##### Comparison for raw-int entries

The ACS entry value stored under `measurement-values-map` codepoint 15 is a raw int value, which must have type `raw-int-type-choice`.

> Consider an `int` ACS entry value named ENTRY in a `measurement-values-map` codepoint (e.g., 15) that allows comparing `int` against a either another `int` or an `int-range` named CONDITION.

Reviewer 2: ```suggestion
Consider an `int` ACS entry value named ENTRY in a `measurement-values-map` codepoint (e.g., 15) that allows comparing `int` against either another `int` or an `int-range` named CONDITION.
```

> Consider an `integral` ACS entry value named ENTRY in a `measurement-values-map` codepoint (e.g., 15) that allows comparing `integral` against a either another `integral` or an `integral-range` named CONDITION.

Reviewer 2: ```suggestion
Consider an `integral` ACS entry value named ENTRY in a `measurement-values-map` codepoint (i.e., 15) that allows comparing `integral` against either another `integral` or an `integral-range` named CONDITION.
```

Author: This is not the intention for int-range. Any entry that chooses to use tagged-int-range (i.e., "allows comparing") is to use these comparison rules.

*  If CONDITION is an `integral` then an equality comparison is performed with ENTRY. Equality between major type 0 and unsigned bignum or conversely major type 1 and negative bignum first requires translation of small ints to bignum to compare.

> *  If CONDITION is an `integral-range` tagged with #6.564, then a range inclusion comparison is performed.

Reviewer 2: ```suggestion
*  If CONDITION is an `integral-range` (CBOR tag 564), then a range inclusion comparison is performed.
```

The comparison MUST return true if and only if all the following conditions are true:

>     + CONDITION.min is `null` or ENTRY is greater than or equal to the CONDITION.min.

Reviewer 2: ```suggestion
    + CONDITION.min is `null` or ENTRY is greater than or equal to CONDITION.min.
```

    + CONDITION.max is `null` or ENTRY is less than or equal to CONDITION.max.

> Consider an `integral-range` or `integral-range` tagged with #6.564 value named ENTRY in a `measurement-values-map` codepoint (e.g., 15) that allows comparing an `int-range` against either another `int-range` or an `int` named CONDITION.

Reviewer 2: ```suggestion
Consider an `integral` or `integral-range` (CBOR tag 564) value named ENTRY in a `measurement-values-map` codepoint (i.e., 15) that allows comparing an `integral-range` against either another `integral-range` or an `integral` named CONDITION.
```

Reviewer 3: Is the justification for changing 'int' to 'integral' because the numeric is defined in terms of int, uint, and big integers?

Author: Will remove bigint and just say raw-int for simplicity.

*  If CONDITION is an `integral`, then the comparison MUST return true if and only if ENTRY.min and ENTRY.max are both equal to CONDITION.

> *  If CONDITION is an `integral-range` tagged with #6.564, then a range subsumption comparison is performed (i.e., the condition range includes all values of the entry range).

Reviewer 2: ```suggestion
*  If CONDITION is an `integral-range` (CBOR tag 564), then a range subsumption comparison is performed (i.e., the condition range includes all values of the entry range).
```
